tfhe/src/core_crypto/algorithms/lwe_keyswitch.rs

```diff
@@ -315,6 +316,137 @@ pub fn keyswitch_lwe_ciphertext_other_mod<Scalar, KSKCont, InputCont, OutputCont
     }
 }
 
+/// Keyswitch an [`LWE ciphertext`](`LweCiphertext`) with a certain InputScalar type to represent
+/// data encrypted under an [`LWE secret key`](`LweSecretKey`) to another [`LWE secret
+/// key`](`LweSecretKey`) using a different OutputScalar type to represent data.
+///
+/// # Notes
+///
+/// This function only supports power of 2 moduli and going from a large InputScalar with
+/// `input_bits` to a a smaller OutputScalar with `output_bits` and `output_bits` < `input_bits`.
+///
+/// The product of the `lwe_keyswitch_key`'s
+/// [`DecompositionBaseLog`](`crate::core_crypto::commons::parameters::DecompositionBaseLog`) and
+/// [`DecompositionLevelCount`](`crate::core_crypto::commons::parameters::DecompositionLevelCount`)
+/// needs to be smaller than `output_bits`.
+pub fn keyswitch_lwe_ciphertext_with_scalar_change<
+    InputScalar,
+    OutputScalar,
+    KSKCont,
+    InputCont,
+    OutputCont,
+>(
+    lwe_keyswitch_key: &LweKeyswitchKey<KSKCont>,
+    input_lwe_ciphertext: &LweCiphertext<InputCont>,
+    output_lwe_ciphertext: &mut LweCiphertext<OutputCont>,
+) where
+    InputScalar: UnsignedInteger,
+    OutputScalar: UnsignedInteger + CastFrom<InputScalar>,
+    KSKCont: Container<Element = OutputScalar>,
+    InputCont: Container<Element = InputScalar>,
+    OutputCont: ContainerMut<Element = OutputScalar>,
+{
+    assert!(
+        InputScalar::BITS > OutputScalar::BITS,
+        "This operation only supports going from a large InputScalar type \
+        to a strictly smaller OutputScalar type."
+    );
+    assert!(
+        lwe_keyswitch_key.decomposition_base_log().0
+            * lwe_keyswitch_key.decomposition_level_count().0
+            <= OutputScalar::BITS,
```

Comment:
I think this should be the `InputScalar` here

Reply:
if it holds for the OutputScalar which is smaller than InputScalar then it also holds for the InputScalar, and it should hold for both :)

Reply:
Indeed, but it's more restrictive than it could be I think
Why should it hold for both?

Reply:
The current decomposer formulas only work if level * base_log < Bits, the decomposer does an approximation over level * base_log bits, so it means that we need the approximated bits to fit in the OutputScalar so approximated bits counts <= OutputScalar::BITS